Blog post on the version 1-3 upgrade (#5743)
unlock-protocol/unlock: unlock-protocol.com/blog/emergency-upgrade-1-3.md

---
title: Emergency Upgrade
subTitle: Version 1-3 is now available
authorName: Nick Mancuso
publishDate: January 17, 2020
description: Explaining what went wrong and why we released a new version so quickly after our last upgrade.
image: /static/images/blog/unlocking-smart-contracts/code.jpeg
---

> We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our most recent smart-contract update.

Reviewer: ```suggestion
We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our [most recent smart-contract update](https://unlock-protocol.com/blog/version-1-2/).
```

Reviewer: Just adding a link ;)

There were problems with this release which if targeted by an attacker had the potential to do one of the following:
 - Steal money made from previous key sales.
 - Break every lock (and key) using the new version - blocking new sales, preventing key owners from being recognized and preventing the lock owner from accessing any funds previously made.

> We fixed the second issue already, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the lock owners themselves to upgrade ASAP.

Reviewer: it seems to imply that the first issue not fixed?

```suggestion
We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the lock owners themselves to upgrade ASAP.
```

> This only impacts locks created in the past couple weeks. If you recently created a lock, please go to the Unlock dashboard to check if you are impacted.  The dashboard will flag any impacted lock, asking you to upgrade.

Reviewer: Let's be more precise on dates (ideally even block numbers!)

Reviewer: And lets something along the lines of "if the dashboard does not show a warning  like the one below, your lock is safe". 

[I'll send a srceenshot ASAP]

Reviewer: <img width="1293" alt="image" src="https://user-images.githubusercontent.com/17735/72651244-cb591000-3950-11ea-8cc0-8b08b1c1224d.png">

Author: Will do.  
Issue introduced: https://etherscan.io/tx/0x159252de7b715cb7e773e5eb54435cd19a8563a574cc5ecfd4f6fcff7bc7204f
Issue resolved: https://etherscan.io/tx/0xa66dc730133db01cc39a30ea7f968558e6f4e5aa1305c5d9851e0547810f71c8

So what happened?  Let’s start with the issue we were able to fix right away and then we’ll get to the juicy stuff.

## Break every lock using the new version

> This issue is about our deployment process.  We simply needed to take action before an attacker.  We will avoid this race condition in the future.

We recently changed the locks deployed to use the minimal proxy standard, [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167).  This significantly reduces the cost to create new locks and leads to less blockchain bloat.  When using proxies any data set in the constructor is lost so instead we call a single-use function called `initialize`.

During normal operation a user calls `createLock` on our main Unlock contract.  That function will deploy the lock and call `initialize` in the same transaction.  It’s responsible for recording data such as the lock owner and the key price.

Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.

We forgot to `initialize` the template!

> If someone figured this out before we did they could have called `initialize` on the template to make themselves the owner and then make an owner-only call to `selfdestruct` the contract.  

Reviewer 2: I wonder if we should mention why we chose to keep `selfdestruct` in the first place? Minor thing, but it was a conscious decision with the goal of helping minimize bloat on the blockchain.

Reviewer 2: openzeppelin even has a warning about this here: https://docs.openzeppelin.com/upgrades/2.6/writing-upgradeable#potentially-unsafe-operations

Reviewer: I think it's worth adding this link to the blog post!

Author: Added both, thanks

Once the template has been destroyed, all proxies attempting to read the implementation logic will fail. This means no new purchases, no reading who owns a key, and no way to withdraw any funds raised from previous key sales.

> We initialized the template and renounced ownership so that no one can ever call selfdestruct. Going forward we have added this to our deployment checklist, ensuring the template is initialized before Unlock is upgraded so that there is no window where someone could do damage by claiming ownership first.

Reviewer 2: could we add these steps (initialize and renounceOwnership) to the deploy-template script for a more robust fix here? I personally just deployed the template using myCrypto, but would be happy to switch to using a script, especially if we automated these actions. Thenwe could say "Going forward we have added this to our deployment script" which in my mind instills more confidence.

Reviewer: It's actually wip ;)

Reviewer: Lets finish a very bold and re-assuring statement saying that this issue has now been fixed and that no one wa able to "abuse" it in any way.

Author: Added both.

## Steal money from key sales

> > This issue is due to a bug in our code.  The bug has been fixed so that any new locks created going forward are fine.  But some locks were impacted and we need lock owners to do the upgrade.

Reviewer: Great!

When you inherit multiple contracts that require initialization, the top-level contract implements `initialize` and then forwards relevant information to the others.

There are three ways this can be done safely:
 1) Ensure that only one `initialize` function is public or external.
 2) Add a modifier to each `initialize` function to ensure it cannot be called twice.
 3) Use the same function signature so that only one implementation is accessible, and ensure that function cannot be called twice.

(1) is not a complete solution for us since we depend on OpenZeppelin.  In order for them to support a variety of use cases every contract uses a public initialize function.  They use option (2) instead.  Option (3) is a little risky as we make changes and add more features.

There were 2 `initialize` calls for which we did not handle this correctly:

```
  function initialize(
    address _tokenAddress
  ) public
  {
    tokenAddress = _tokenAddress;
    require(
      _tokenAddress == address(0) || IERC20(_tokenAddress).totalSupply() > 0,
      'INVALID_TOKEN'
    );
  }
```

and

```
  function initialize() public
  {
    // default to 10%
    refundPenaltyBasisPoints = 1000;
  }
```

> These two calls are public and have no modifier.  This means anyone can call them at any time.

Reviewer: ```suggestion
These two calls are public and have no modifier.  This means anyone can call them at any time on affected locks.
```

---
title: Emergency Upgrade
subTitle: Version 1-3 is now available
authorName: Nick Mancuso
publishDate: January 17, 2020
description: Explaining what went wrong and why we released a new version so quickly after our last upgrade.
image: /static/images/blog/emergency-upgrade-1-3/code_vulnerability.jpg
---

We recently noticed a vulnerability which if abused could have had bad consequences – but we caught it early and fixed it promptly. This post aims to explain the errors we made with our [most recent smart-contract update](https://unlock-protocol.com/blog/version-1-2/).

There were problems with this release which if targeted by an attacker had the potential to do one of the following:
 - Steal money made from previous key sales.
 - Break every lock (and key) using the new version - blocking new sales, preventing key owners from being recognized and preventing the lock owner from accessing any funds previously made.

> We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the owners of these locks themselves to upgrade ASAP. (there are 14 locks which affected)

Reviewer: ```suggestion
We fixed the second issue very quickly, so no worries there.  But funds are at risk if your lock is impacted.  For that we need the owners of these locks themselves to upgrade ASAP. (there are 14 locks which are affected)
```

> This only impacts locks created between Jan 6th until Jan 17th (locks created in blocks #9227829 - #9295958). If you recently created a lock, please go to the Unlock dashboard to check if you are impacted. If the dashboard does not show a warning like the one below, your lock is safe.

Reviewer: ```suggestion
This only impacts locks created between Jan 6th until Jan 17th (locks created between the block #9227829 and the block #9295958). If you recently created a lock, please go to the Unlock dashboard to check if you are impacted. If the dashboard does not show a warning like the one below, **your lock is safe**.
```

<p style="text-align:center">
	<img src="/static/images/blog/emergency-upgrade-1-3/dashboard_warning.png" width="750px" alt="Dashboard warning">
</p>

So what happened?  Let’s start with the issue we were able to fix right away and then we’ll get to the juicy stuff.

## Break every lock using the new version

> This issue is about our deployment process.  We simply needed to take action before an attacker.  We will avoid this race condition in the future.

We recently changed the locks deployed to use the minimal proxy standard, [EIP-1167](https://eips.ethereum.org/EIPS/eip-1167).  This significantly reduces the cost to create new locks and leads to less blockchain bloat.  When using proxies any data set in the constructor is lost so instead we call a single-use function called `initialize`.

During normal operation a user calls `createLock` on our main Unlock contract.  That function will deploy the lock and call `initialize` in the same transaction.  It’s responsible for recording data such as the lock owner and the key price.

> Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock proxies to leverage.

Reviewer: ```suggestion
Proxies work by reading the implementation logic from a template.  The template is nothing more than a contract containing the logic / source code for the lock implementations to leverage.
```

Author: hmm that sounds a little confusing to me.  maybe `... / source code for each individual lock to leverage`

Author: I pushed with the text I suggested.  let me know if this sounds alright.
